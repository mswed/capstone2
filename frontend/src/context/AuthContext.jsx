import React, { useState, useEffect } from 'react';
import { jwtDecode } from 'jwt-decode';
import GrumpyApi from '../services/api';

// We create the context
const AuthContext = React.createContext();

/**
 * This is a logic only component that wraps our entire app and sets the
 * authentication context and its functions
 *
 * @param {Array}  children - Components to wrap
 * @returns {AuthContext.Provider} - with auth functions for easy access
 */

const AuthProvider = ({ children }) => {
  // We grab stuff from local storage if available
  const storedToken = localStorage.getItem('grumpy-token');
  const storedUser = localStorage.getItem('grumpy-user');

  // We set the token right away so we don't run in to authorization issues
  GrumpyApi.token = storedToken || null;

  // Initialize state
  const [token, setToken] = useState(storedToken);
  const [currentUser, setCurrentuser] = useState(storedUser);

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    if (token && token.trim !== '') {
      // We only set the token if it's not empty

      GrumpyApi.token = token;
      localStorage.setItem('grumpy-token', token || '');
      localStorage.setItem('grumpy-user', currentUser || '');
    } else {
      GrumpyApi.token = null;
      localStorage.removeItem('grumpy-token');
      localStorage.removeItem('grumpy-user');
    }
  }, [token]);

  // TODO:
  // // We also want to get the user favorites here
  // useEffect(() => {
  //   if (!token) return;
  //   async function getUserData() {
  //     const userData = await GrumpuyApi.getUser(currentUser);
  //     setFavorites(new Set(userData.favorites));
  //   }
  //   getUserData();
  // }, [token]);

  const login = async (username, password) => {
    try {
      const authToken = await GrumpyApi.login(username, password);
      setToken(authToken);
      const decoded = jwtDecode(authToken);
      setCurrentuser(decoded.username);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await GrumpyApi.register(username, password, firstName, lastName, email);
      logout();
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = () => {
    // TODO: Do I need to logout of the backend here too?
    setToken('');
  };

  // TODO: We want to add favorites here
  // const addFavorite = async (jobId) => {
  //   const res = await GrumpuyApi.favoriteFormat(currentUser, jobId);
  //   if (res) {
  //     setFavorites((currentFavorites) => currentFavorites.add(res.favorites));
  //   }
  // };
  //

  // TODO: get items that have already been favorited
  // const alreadyFavorited = (formatId) => {
  //   return favFormats.has(formatId);
  // };

  // Create the value object
  const value = {
    token,
    currentUser,
    login,
    register,
    logout,
  };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
