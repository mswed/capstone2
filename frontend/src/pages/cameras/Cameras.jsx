import { useState, useEffect } from 'react';
import { Container } from 'react-bootstrap';
import GrumpyApi from '../../services/api';
import Loading from '../../components/ui/Loading';
import RemoteSearchForm from '../../components/forms/RemoteSearchForm';
import CameraList from '../../features/cameras/components/CameraList';

const Cameras = () => {
  const [cameras, setCameras] = useState([]);
  const [searchTerm, setSearchTerm] = useState('');
  // We need to check if we're still loading so we won't run into trying to
  // render a null state
  const [isLoading, setIsLoading] = useState(true);

  /**
   * Get all the cameras in the database. Eventually this should have a limit
   * or pagination but for now we get everything. This is called by useEffect
   * if the searchTerm changes and is empty
   *
   * @returns {Array} all cameras in the database
   */

  const getAllCameras = async () => {
    try {
      const response = await GrumpyApi.getCameras();
      setCameras(response);
    } catch (error) {
      console.error('Error fetching cameras', error);
    } finally {
      setIsLoading(false);
    }
  };

  /**
   * Get found cameras in the databse. This is called by useEffect
   * if the searchTerm changes and is NOT empty
   *
   * @returns {Array} found cameras in the database
   */

  const findCameras = async () => {
    try {
      const response = await GrumpyApi.findCameras(searchTerm);
      setCameras(response);
    } catch (error) {
      console.error('Error fetching cameras', error);
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    if (searchTerm.trim() !== '') {
      findCameras();
    } else {
      getAllCameras();
    }
  }, [searchTerm]);

  if (isLoading) {
    return <Loading />;
  }
  return (
    <Container>
      <RemoteSearchForm search={setSearchTerm} />
      <CameraList cameras={cameras} showMake={true} />
    </Container>
  );
};

export default Cameras;
