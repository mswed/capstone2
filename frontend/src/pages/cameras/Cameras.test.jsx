import { vi, it, expect, beforeEach } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Cameras from './Cameras.jsx';
import { AuthContext } from '../../context/AuthContext.jsx';
import { Component } from 'react';

// Mock API calls
vi.mock('../../services/api', () => ({
  default: {
    getCameras: vi.fn(() =>
      Promise.resolve([
        {
          id: 1,
          make: 1,
          makeName: 'ARRI',
          model: 'Alexa 35',
          sensorType: 'Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array',
          sensorSize: 'Super 35',
          maxFilmbackWidth: 27.99,
          maxFilmbackHeight: 19.22,
          maxImageWidth: 4608,
          maxImageHeight: 3164,
          minFrameRate: 0.75,
          maxFrameRate: 120,
          notes: 'sample notes',
        },
      ])
    ),
    findCameras: vi.fn(() =>
      Promise.resolve([
        {
          id: 2,
          make: 2,
          makeName: 'RED',
          model: 'KOMODO-X',
          sensorType: 'KOMODO-Xâ„¢ 19.9MP Super 35mm Global Shutter CMOS',
          sensorSize: 'Super 35',
          maxFilmbackWidth: 27.03,
          maxFilmbackHeight: 14.26,
          maxImageWidth: 6144,
          maxImageHeight: 3240,
          minFrameRate: 24,
          maxFrameRate: 240,
        },
      ])
    ),
  },
}));

describe('Cameras component', () => {
  /**
   * Render the componenet with authentication context
   *
   * @returns {Component} - Makes component wrapped in MemoryRouter and AuthContext
   */

  const renderWithAuth = (authValue = { token: 'fake-token', currentUser: 'testUser' }) => {
    return render(
      <MemoryRouter>
        <AuthContext.Provider value={authValue}>
          <Cameras />
        </AuthContext.Provider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    // Clear mocks
    vi.clearAllMocks();
  });

  it('renders without crashing', async () => {
    renderWithAuth();
    await screen.findByText('ARRI');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderWithAuth();

    await screen.findByText('ARRI');
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct info', async () => {
    renderWithAuth();

    await waitFor(() => {
      // Table headers
      expect(screen.getByText('Image')).toBeInTheDocument();
      expect(screen.getByText('Make')).toBeInTheDocument();
      expect(screen.getByText('Model')).toBeInTheDocument();
      expect(screen.getByText('Sensor Type')).toBeInTheDocument();
      expect(screen.getByText('Max Filmback Size')).toBeInTheDocument();
      expect(screen.getByText('Max Resolution')).toBeInTheDocument();

      // Table data
      expect(screen.getByText('ARRI')).toBeInTheDocument();
      expect(screen.getByText('Alexa 35')).toBeInTheDocument();
      expect(screen.getByText('Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array')).toBeInTheDocument();
      expect(screen.getByText('27.99mm x 19.22mm')).toBeInTheDocument();
      expect(screen.getByText('4608 x 3164')).toBeInTheDocument();
    });
  });

  it('can search for cameras', async () => {
    renderWithAuth();

    await screen.findByText('ARRI');

    // Fill out the required fields
    const searchInput = screen.getByPlaceholderText('Enter search term...');
    const searchButton = screen.getByRole('button', { name: /search/i });

    fireEvent.change(searchInput, { target: { value: 'RED' } });
    fireEvent.click(searchButton);

    await waitFor(() => {
      // Table headers
      expect(screen.getByText('Image')).toBeInTheDocument();
      expect(screen.getByText('Make')).toBeInTheDocument();
      expect(screen.getByText('Model')).toBeInTheDocument();
      expect(screen.getByText('Sensor Type')).toBeInTheDocument();
      expect(screen.getByText('Max Filmback Size')).toBeInTheDocument();
      expect(screen.getByText('Max Resolution')).toBeInTheDocument();

      // Table data
      expect(screen.getByText('RED')).toBeInTheDocument();
    });
  });
});
