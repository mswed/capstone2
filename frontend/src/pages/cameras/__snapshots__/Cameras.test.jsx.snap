// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cameras component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class=""
    >
      <div
        class="my-3 input-group"
      >
        <input
          class="form-control"
          placeholder="Enter search term..."
          value=""
        />
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search!
        </button>
      </div>
    </form>
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-hover"
      >
        <thead>
          <tr>
            <th>
              Image
            </th>
            <th>
              Make
            </th>
            <th>
              Model
            </th>
            <th>
              Sensor Type
            </th>
            <th>
              Max Filmback Size
            </th>
            <th>
              Max Resolution
            </th>
            <th>
              Action
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <img
                alt="Image of Alexa 35"
                src="/media/camera_images/missing_image.png"
                style="width: 60px;"
              />
            </td>
            <td>
              ARRI
            </td>
            <td>
              Alexa 35
            </td>
            <td>
              Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array
            </td>
            <td>
              27.99mm x 19.22mm
            </td>
            <td>
              4608 x 3164
            </td>
            <td>
              <a
                class="btn btn-outline-primary btn-sm"
                data-discover="true"
                href="/cameras/1"
              >
                View
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
